crates/circuit/src/dag_circuit.rs

```diff
@@ -2618,8 +2545,18 @@ def _format(operand):
                 [NodeType::ClbitIn(bit1), NodeType::ClbitIn(bit2)] => Ok(bit1 == bit2),
                 [NodeType::QubitOut(bit1), NodeType::QubitOut(bit2)] => Ok(bit1 == bit2),
                 [NodeType::ClbitOut(bit1), NodeType::ClbitOut(bit2)] => Ok(bit1 == bit2),
-                [NodeType::VarIn(var1), NodeType::VarIn(var2)] => var1.bind(py).eq(var2),
-                [NodeType::VarOut(var1), NodeType::VarOut(var2)] => var1.bind(py).eq(var2),
+                [NodeType::VarIn(var1), NodeType::VarIn(var2)] => self
+                    .vars
+                    .get(*var1)
+                    .unwrap()
+                    .bind(py)
+                    .eq(other.vars.get(*var2).unwrap()),
+                [NodeType::VarOut(var1), NodeType::VarOut(var2)] => self
+                    .vars
+                    .get(*var1)
+                    .unwrap()
+                    .bind(py)
+                    .eq(other.vars.get(*var2).unwrap()),
```

Comment:
Maybe I'm not getting this right so feel free to correct me, the comparison here is not done by index because `Var` objects are not opaque. Meaning that they are different fundamentally speaking and are not differentiated by index only. Is this why we're comparing them this way here?

Reply:
Pretty much, though I'd say they _are_ opaque in that they're nothing more than keys that users must exchange with the `DAGCircuit` to get information about the actual variable.

The difference between `Var` and `{Qu,Cl}bit` is pretty much that a variable's index doesn't mean _anything_. With bits, the index is always also the position within the DAG's bit list, be it qubits or clbits, so when comparing two DAGs it's enough to just compare the indices. A `Var`'s index doesn't mean anything because the relationship between `Var` and our DAG is that a DAG has a *set* of `Var`s, not a list. We just happen to be storing them in a list-like structure (`BitData`).

I'm glad you mentioned this, because I think it really means we ought to make sure that `Var` isn't `Eq` or `Ord` outside of `DAGCircuit`, since it doesn't actually have those semantics. The right thing to do is probably to make both `Wire` and `Var` private to `DAGCircuit` and if we must expose a wire-like enum, have it be an enumeration of `Qubit`, `Clbit`, and `PyObject` like it is now, and convert to it within `DAGCircuit` from the internal `Wire` enum of `Qubit`, `Clbit`, `Var` at the boundary.

I'll look at the feasibility of doing this as part of this PR and report back.